src/main/java/net/troja/eve/esi/auth/OAuth.java

```diff
@@ -26,25 +26,30 @@
     private static final String URI_AUTHENTICATION = URI_OAUTH + "/authorize";
     private static final String URI_ACCESS_TOKEN = URI_OAUTH + "/token";
 
-    private String accessToken;
-    private long validUntil;
     private String refreshToken;
     private String clientId;
     private String clientSecret;
     private OAuth2CodeGrantFlow oAuthFlow;
-    private final Map<String, AccessTokenData> accessTokenCache = new HashMap<>();
+    private static final Object STATIC_LOCK = new Object();
+    private static final Map<String, AccessTokenData> ACCESS_TOKEN_CACHE = new ConcurrentHashMap<>();
 
     @Override
     public void applyToParams(final List<Pair> queryParams, final Map<String, String> headerParams) {
-        if (refreshToken != null && validUntil < System.currentTimeMillis()) {
-            try {
-                refreshToken();
-            } catch (final ProcessingException ex) {
-                // This error will be handled by ESI once the request is made
+        //Check if we need a new access token
+        synchronized(STATIC_LOCK) { //This block is synchronized across all threads - so we don't update the access token more than once
+            AccessTokenData accessTokenData = ACCESS_TOKEN_CACHE.get(getAuthKey());
+            if (refreshToken != null  && (accessTokenData == null || accessTokenData.getValidUntil() < System.currentTimeMillis())) {
+                try {
+                    refreshToken();
+                } catch (final ProcessingException ex) {
+                    // This error will be handled by ESI once the request is made
+                }
```

Comment:
Suggest you move this block after line 50 below and move the synchronization inside the refreshToken try block to reduce contention.  So starting from line 50 it would look like this (this works as long as you always REPLACE AccessTokenData, and not update it in place):

AccessTokenData accessTokenData = ...
if (refreshToken != null && (accessTokenData == null ...
  synchronized (STATIC_LOCK) {
    // Have to check the token again in case someone beat you to the update
    accessTokenData = ACCESS_TOKEN_CACHE.get(...
    if (accessTokenData == null ...) {
      try {
        refreshToken();
      } catch(...
      }
      // We updated the token, so we need to re-read
      accessTokenData = ACCESS_TOKEN_CACHE.get(...
    }
  }